src/check_executor.rs

```diff
@@ -0,0 +1,169 @@
+
+            async move {
+                let ScheduledCheck {
+                    config,
+                    tick,
+                    resolve_tx,
+                } = scheduled_check;
+
+                let _ = tokio::spawn(async move {
+                    let check_result = job_checker.check_url(&config, &tick).await;
+
+                    if let Err(e) = job_producer.produce_checker_result(&check_result) {
+                        tracing::error!(error = ?e, "Failed to produce check result");
+                    }
+
+                    record_result_metrics(&check_result);
+                    tracing::info!(result = ?check_result, "Check complete");
+
+                    resolve_tx
+                        .send(check_result)
+                        .expect("Failed to resolve completed check");
+                })
+                .await;
+            }
+        })
+        .await;
+}
+
+fn record_result_metrics(result: &CheckResult) {
+    // Record metrics
+    let CheckResult {
+        status,
+        scheduled_check_time,
+        actual_check_time,
+        duration,
+        status_reason,
+        ..
+    } = result;
+
+    let status_label = match status {
+        CheckStatus::Success => "success",
+        CheckStatus::Failure => "failure",
+        CheckStatus::MissedWindow => "missed_window",
+    };
+    let failure_reason = match status_reason.as_ref().map(|r| r.status_type) {
+        Some(CheckStatusReasonType::Failure) => Some("failure"),
+        Some(CheckStatusReasonType::DnsError) => Some("dns_error"),
+        Some(CheckStatusReasonType::Timeout) => Some("timeout"),
+        None => None,
+    };
+
+    // Record duration of check
+    if let Some(duration) = duration {
+        metrics::histogram!(
+            "check_result.duration_ms",
+            "histogram" => "timer",
+            "status" => status_label,
+            "failure_reason" => failure_reason.unwrap_or("ok"),
+        )
+        .record(duration.num_milliseconds() as f64);
+    }
+
+    // Record time between scheduled and actual check
+    metrics::histogram!(
+        "check_result.delay_ms",
+        "histogram" => "timer",
+        "status" => status_label,
+        "failure_reason" => failure_reason.unwrap_or("ok"),
+    )
+    .record((*actual_check_time - *scheduled_check_time).num_milliseconds() as f64);
+
+    // Record status of the check
+    metrics::counter!(
+        "check_result.processed",
+        "status" => status_label,
+        "failure_reason" => failure_reason.unwrap_or("ok"),
+    )
+    .increment(1);
+}
+
+#[cfg(test)]
+mod tests {
+    use chrono::Utc;
+    use similar_asserts::assert_eq;
+    use tokio::sync::oneshot;
+    use uuid::Uuid;
+
+    use super::*;
+    use crate::{
+        checker::dummy_checker::DummyChecker, producer::dummy_producer::DummyResultsProducer,
+    };
+
+    #[tokio::test]
+    async fn test_executor_simple() {
```

Comment:
Worth having a test where the producer errors?

Reply:
Yeah, though I'm still wondering what it should actually do.

Reply:
Could be something we handle as a separate task. I think that the produce here isn't actually a guarantee that we wrote to Kafka anyway, there's a callback for that from what I remember

Reply:
Yeah let's handle this separately I think.